src/main/java/org/openstreetmap/atlas/geography/atlas/raw/sectioning/NodeOccurrenceCounter.java

```diff
@@ -0,0 +1,93 @@
+package org.openstreetmap.atlas.geography.atlas.raw.sectioning;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import org.openstreetmap.atlas.geography.Location;
+import org.openstreetmap.atlas.geography.atlas.items.Edge;
+import org.openstreetmap.atlas.geography.atlas.raw.slicing.temporary.TemporaryNode;
+
+/**
+ * Keeps track of all {@link TemporaryNode}s created for an atlas {@link Edge} and the occurrence of
+ * it in the underlying polyline. This is used specifically for raw atlas way-sectioning.
+ *
+ * @author mgostintsev
+ */
+public class NodeOccurrenceCounter
+{
+    // Mapping of node to occurrence
+    private final Map<TemporaryNode, Integer> nodes = new HashMap<>();
+
+    /**
+     * Adds the given {@link TemporaryNode} to the store
+     *
+     * @param node
+     *            The {@link TemporaryNode} to add
+     */
+    public void addNode(final TemporaryNode node)
+    {
+        this.nodes.put(node, 1);
+    }
+
+    /**
+     * Get the {@link TemporaryNode} at this {@link Location}.
+     *
+     * @param location
+     *            The target {@link Location}
+     * @return an {@link Optional} containing the {@link TemporaryNode}
+     */
+    public Optional<TemporaryNode> getNode(final Location location)
+    {
+        for (final TemporaryNode node : this.nodes.keySet())
+        {
+            if (node.getLocation().equals(location))
```

Comment:
do we know how many nodes are typically expected in a changeset?  Would it more performant to track this as <Location,Int> instead and use atlas.itemsAt() calls to get the node?

Reply:
if we convert this to use streams, would parallelism help here?

Reply:
1. This code is for Nodes per Line-To-Become-Edge, so number of nodes will be significantly less than if looking at an entire changeset. Number of nodes is [1-999]. A closed-loop and a way with 1000 intersections. Technically an OSM Way can have 2000 shape-points, but since we're sectioning, we're limited to 999 for the identifiers to work with. Most ways don't have nearly this much.
2. Can you explain more on tracking as <Location, Int>? That's essentially what a TemporaryNode is - it has the node identifier and location (bare minimum to identify a specific node). The goal is to query by Location. I thought about using the atlas.pointsAt(Location) similar to addPointToNodeList call - https://github.com/osmlab/atlas/blob/18929714472a6c35fea9e046254ca810cb0415f1/src/main/java/org/openstreetmap/atlas/geography/atlas/raw/sectioning/WaySectionProcessor.java#L137. This would work, even though this returns all Points at that location, since we built in the restriction that a raw Atlas will have a single point at each location. However, given the small number of Nodes per Edge and to avoid having logic to handle the case of "what if we have two nodes that are very similar or happen to be on top of each other", I thought it was better to have an explicit equality check for each key.
3. I'll try parallelism here, thanks!

Reply:
Did some testing with parallel stream, it actually took longer than the code above. Had more instances of greater than 1ms calls to the method than prior version. Here's the code I used for testing.

return this.getNodes().parallelStream()
                .filter(node -> node.getLocation().equals(location)).findFirst();

Reply:
hmm good point - I was trying to see if iterating through all the nodes was time consuming enough that we might want to save the location -> atlas identifier map to quickly get the node from a given location.